x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java

```diff
@@ -675,4 +686,115 @@ protected void parseCreateField(DocumentParserContext context) throws IOExceptio
+
+            private boolean hasValues;
+
+            ImmediateLeaf(Map<Metric, SortedNumericDocValues> metricDocValues) {
+                this.metricDocValues = metricDocValues;
+            }
+
+            @Override
+            public boolean empty() {
+                return false;
+            }
+
+            @Override
+            public boolean advanceToDoc(int docId) throws IOException {
+                // It is required that all defined metrics must exist. In this case
+                // it is enough to check for the first docValue. However, in the future
+                // we may relax the requirement of all metrics existing. In this case
+                // we should check the doc value for each metric separately
+                dvHasValue = new EnumMap<>(Metric.class);
+                hasValues = false;
+                for (Map.Entry<Metric, SortedNumericDocValues> e : metricDocValues.entrySet()) {
+                    boolean b = e.getValue().advanceExact(docId);
+                    hasValues = hasValues || b;
+                    dvHasValue.put(e.getKey(), b);
+                }
+
+                return hasValues;
+            }
+
+            @Override
+            public void write(XContentBuilder b) throws IOException {
+                if (false == hasValues) {
+                    return;
+                }
+                b.startObject(simpleName);
+                for (Map.Entry<Metric, SortedNumericDocValues> entry : metricDocValues.entrySet()) {
+                    if (dvHasValue.get(entry.getKey())) {
+                        String metricName = entry.getKey().name();
+                        long value = entry.getValue().nextValue();
+                        if (entry.getKey() == Metric.value_count) {
+                            b.field(metricName, value);
+                        } else {
+                            b.field(metricName, NumericUtils.sortableLongToDouble(value));
+                        }
+                    }
+                }
+                b.endObject();
+            }
+        }
+
+        /**
+         * Returns a {@link SortedNumericDocValues} or null if it doesn't have any doc values.
+         * See {@link DocValues#getSortedNumeric} which is *nearly* the same, but it returns
+         * an "empty" implementation if there aren't any doc values. We need to be able to
+         * tell if there aren't any and return our empty leaf source loader.
+         */
+        private SortedNumericDocValues dv(LeafReader reader, Metric metric) throws IOException {
+            String fieldName = subfieldName(name, metric);
+            SortedNumericDocValues dv = reader.getSortedNumericDocValues(fieldName);
+            if (dv != null) {
+                return dv;
+            }
+            NumericDocValues single = reader.getNumericDocValues(fieldName);
+            if (single != null) {
+                return DocValues.singleton(single);
+            }
+            return null;
```

Comment:
Since all metric subfields are stored using the `SortedNumericDocValuesField`, al doc values will be stored as `SortedNumericDocValues`. So, I am not sure  `reader.getNumericDocValues(fieldName)` will ever return non-null values

Reply:
Try removing it! I believe if you store single values it'll flip to `NumericDocValues` anyway.

If you do have to keep loading from both of there it's probably worth making this method `static` and `public` in `NumericFieldMapper` or something like that.